### ISSUE TITLE
How to protect blockchain from *time-warp* attacks?

### ISSUE THREAD

**Issue by nesterow** [amjuarez/bytecoin#202]
Hello,

This issue is related to current _difficulty adjustment algorithm_ and and an attack vector which works as good as DoS for Bytecoin and Monero derivatives, especially for those who does't have established loyal miner community.

### Case
The attacker changes blocks `timestamp` making it most closest to the previous block. In combination with high hash-power (cloud mining services) it causes blocks to solve in short time (2-30 sec). It leads network to raise difficulty. Then attacker leaves network at deadlock for days. Time-warping can be done by adding slight modifications to a mining pool or mining software. Later I'll describe how attacker's profit from it.

### Solutions
Sumocoin were fist who dealt with this problem. They developed a new [_difficulty adjustment algorithm_](https://github.com/sumoprojects/sumokoin/pull/10). Sumocoin's algorithm calculates next difficulty using an average of blockcount (Poisson Distribution), it smooths difficulty and doesn't allow sudden spikes. It is effective when a coin has established miners community, but for new forks the problem stands if they can't overcome hash-power from cloud mining services.

### Live Examples
At least two coins were affected by this attack in the first week of 2018. [Royalties](http://xryexplorer.info/) and [UltraNote](http://explorer.ultranote.org). I helped UltraNote (XUN), so I will talk about them.

On January 2, 2018 XUN network got into deadlock at difficulty 500M. Loyal miners had only 200K/h on official pools and attackers left the network. We learned later that attacker used only ~1M/h and 5-30 second time-warps to put it in deadlock. I've seen the blocks that was solved 1 hour behind and 30 second ahead of the last block. We've estimated that it would take a week to solve next block. So we decided that it is perfect time to apply another difficulty adjustment algorithm and make a hardfork. 

So, we closed repository and applied Sumocoin's algorithm with couple adjustments for our block time (thank you Sumocoin). Hardfork went well and we got out of deadlock, timewarps didn't raise difficulty to enormous numbers and loyal miners started to profit again. It didn't last for long...

Attackers figured that now difficulty would rise smoothly and since they had hash power much larger than our miners can overcome. They started to send us 1,5-3M/h with time-warps essentially stealing money and slowing down the network. No they could't put it in deadlock but since loyal miners can't overcome attacker's hashpower, it essentially works as DoS when attacker leaves the network. So now we have another problem. 

I was afraid to restrict block time on blockchain level when we did the hardfork, because it can slow down network and if network has spike down to low difficulty it can lead to even slower difficulty adjustment. Now i am not sure. 
I did restriction on pool level and later in the daemon RPC, Essentially verifying timestams. But it didn't stop attackers from deploying their own pools.. So the problem stands.

### How does attacker profit?
In our case we were lucky. Because our attacker turned out to be a troll. He opened all cards.
So basically, attackers put new coins to deadlock for 2-5 days. When the coin is new it trades on 1-2 exchanges. So while network is in stall, it's value would raise on exchanges, because deposits are locked and there are no fresh coins. So attackers use coins that they got while attacking. After that the cycle repeats again.
![troll](https://raw.githubusercontent.com/nesterow/trollbase/master/troll1.jpg "troll")

### References for analysis:
[Blochain Expolrer](http://explorer.ultranote.org)
[Attacker's Pool](http://sebseb7.github.io/xun/)

[Sumocoin's Algorithm](https://github.com/sumoprojects/sumokoin/pull/10)

### Big Question
Is it reasonable to restrict block time on blockchain level? For example, rejecting blocks that were solved in less than a minute? 

Thank you.

**Comment by noglance** [comment]
))
вроде не бездельники и могли бы жить, им бы понедельники взять и отменить

**Comment by tomekmarchi** [comment]
Just some two cents. My gut says write an algo that detects the anomalies and simply offers no reward for that block or gives that reward to a random address that is currently mining. In case of DOS attacks best case is either to raise the amount of money it takes to perform the attack or simply ensure the attacker does not benefit from his activities. This would be a quick solution to ensure the network is still moving along but at the same time provides a way for the coins not to make it into the attackers hands. This will at the very least make it worthless for the attacker to carry out his activities.

**Comment by nesterow** [comment]
That's interesting idea. However, no reward would mean no emission. 
I am not sure about sending reward to random miners. I think the network only knows miner's address when miner sends solved block. Maybe it is possible to distribute reward among transactions in the block. But still I think it would make things more complicated.

**Comment by Shinjuku1** [comment]
An algorithm that "detects the anomalies" would be a good start what then just have the attacker's rewards sent right to the genesis address. Any thoughts on that?

**Comment by noglance** [comment]
IPv6...?)

**Comment by aivve** [comment]
KARBOwanec have changed the difficulty algorithm to protect from such attacks using Zawy's formula more than a year ago so your information that it is Sumokoin invented protection is not correct.

**Comment by UltraNote** [comment]
Sorry avivve is that the best solution you have found to contribute for the problem? Currect dates and how found the solution first and implemented.

I beleive in your skills. You know much more than this.

**Comment by nesterow** [comment]
@aivve Thank you. Mistake corrected. Sometimes things work strange - reading different sources I believed they invented it.

**Comment by aivve** [comment]
It's OK, just read through [conversation with author of the algorithm](https://github.com/seredat/karbowanec/commit/231db5270acb2e673a641a1800be910ce345668a#commitcomment-22624474), you might find some ideas. We were [discussing possibilty to punish by reward](https://github.com/seredat/karbowanec/commit/231db5270acb2e673a641a1800be910ce345668a#commitcomment-23983192) too. Here's [the example of the algo that adjusts reward](https://github.com/seredat/karbowanec/commit/231db5270acb2e673a641a1800be910ce345668a#commitcomment-24037066). But it looks like it won't help either, as you can see in further comments. The only thing that comes to my mind is modifying POW algorithm to render it incompatible with Nicehash so attackers won't be able to easily rent massive hashing power. Combined with decent community of miners this can protect from abusing.

**Comment by nesterow** [comment]
Thank you. I hope I'll be able to understand most of it with time and hard work. Zawy12 did very serious work on difficulty algorithms - isn't something that anyone can understand deeply in a short term.

**Comment by nesterow** [comment]
@tomekmarchi Your idea maybe not bad after all. Aivve's leads got me to think about increasing `Reward Penalty` based on time difference compared to the previous block.
Something like that:
```
timespan = current_block.timestamp - tail_block.timestamp
 if (timespan < target time)
   penalty = 1 - timespan / target time
else
  penalty = 0
``` 
I don't know if it is currently possible, but I think it is worth to try.

**Comment by aivve** [comment]
It is possible, you will just need to modify reward calculation function to pass timestamps

**Comment by nesterow** [comment]
I think I'll test it. I have doubts if `Target Time` should be a constant in this case. I think there must be some formula to calculate optimal target time based on prev N blocks. And previous timestamps should be filtered from abnormalities.

**Comment by aivve** [comment]
Timestamps can be manipulated as you described that's why I think constant `target time` should be left intact. Also, decreasing of the reward could slow down emission as Zawy pointed out, though I don't get it since reward formula should compensate.

**Comment by nesterow** [comment]
Yes. I agree. A simple solution is usually most effective. The reason why I think about dynamic `target time` is that there's no guarantee that a legit miner finds a block in less time. I was assuming that there is a way to identify abnormalities in previous blocks, but I came to understanding that there would be more ways to cheat on it.
I think reward penalty is taken in account in emission logic, I need to check how to be sure if it would compensate penalties. And even If emission slows down for a bit - I don't consider it a bad thing. From economist's perspective it can provide more value on the market. Although from miner's perspective it can look as a bad thing.

**Comment by nesterow** [comment]
Yes, the emission logic compensates penalties. Base Reward function takes in account already generated coins and block height. The reward should raise within reward increase interval taking in account already generated coins. Only possible problem I see is that if there are too much penalties at some point the reward can overcome `MAX_BLOCK_REWARD` and emission would go under supply.  
```cpp
uint64_t Currency::baseRewardFunction(uint64_t alreadyGeneratedCoins, uint32_t height) const {
  if (height == 1) return ICO_BLOCK_REWARD;
  
  uint64_t incrIntervals = static_cast<uint64_t>(height) / REWARD_INCREASE_INTERVAL;
  assert(incrIntervals < POWERS_OF_TEN.size());
  
  uint64_t base_reward = START_BLOCK_REWARD * POWERS_OF_TEN[incrIntervals];	
  base_reward = (std::min)(base_reward, MAX_BLOCK_REWARD);
  base_reward = (std::min)(base_reward, m_moneySupply - alreadyGeneratedCoins);
  
  return base_reward;
}
```

**Comment by FndNur1Labs** [comment]
is there PR for this?

**Comment by nesterow** [comment]
@FndNur1Labs No, as I pointed in my fix to this comment - this particular function doesn't take in account already generated coins until final turn. As far as I understand, it won't go under supply because of `MIN(base_reward, moneySupply - alreadyGeneratedCoins)`. I am going to do tests with different reward functions to see what happens with emission when there are big reward penalties.

**Comment by tomekmarchi** [comment]
Let us take look at this from a macro perspective. We are talking about heavy market regulation. In particular unfair competition laws. The way it's handled in reality is to instead reward the little guy(s) to encourage more options and more competition in the market long term. With that being said we can look at it from an economic perspective and address the problem. Similar to the US; by making a monopoly effectively illegal or penalized. If someone has a monopoly on the hashrate aka unfair competition then we redirect and or spread out the funds as much as possible to addresses from the transactions that are in previous blocks. The transactions can match some specific criteria for example some that seem to be representative of the poor, middle, and upper middle class. This wouldn't slow down emission rate as it would just be pushing funds to everyone else. Ensuring effectively that we still reward the capitalist who invested time and money into purchasing large hashpower. However, we also set some aside to reward others who have made transactions on the network that are of the poor, middle, and upper middle class. From the looks of this, it seems to be an economic issue purely and needs to be addressed as such. This can be viewed as a sort of taxation on having a monopoly on the market while still ensuring that the market as a whole sees growth. Just some more two cents.

**Comment by FndNur1Labs** [comment]
why not make sub-chain system for secure. mean main chain with
log[temporary] can only for client used. and codes protected by self. or
used some engine for better supply chain secured?

**Comment by aejontargaryen** [comment]
i like @tomekmarchi approach of redistributing block reward in the case of a blatant monopolistic attack. I feel like the mined money unlock window could possibly serve a strong function somehow, too.
also @nesterow  would -block grant full reward zone- share this same principle of emission

**Comment by sebseb7** [comment]
I was the "attacker" and you got your Story completely wrong. Why did nobody ask me what really happened? lol 

keep looking... :-D

**Comment by nesterow** [comment]
@aejontargaryen Yes, I believe it would share this principle if the reward function has compensation on penalized amount, it wouldn't  really matter how it was penalized.

**Comment by sebseb7** [comment]
"Is it reasonable to restrict block time on blockchain level? For example, rejecting blocks that were solved in less than a minute?"

You cannot enforce that. The undocumented hardfork you did could have easlity destoryed xun and it made a lot of problems. Also: removing the code form public view only worsened the situation. Xun is a failure. The devs are incompetent and criminal.

**Comment by sebseb7** [comment]
"You can delay block template refresh making miners to solve the blocks with most closest interval to prevoius block. "

That only makes it much more profitable to the "attacker"  - thank you !

**Comment by nesterow** [comment]
@sebseb7 This is the story from my perspective. I would really want to hear it from your perspective. The code was open a week ago. The hardfork we did was necessary because legacy difficulty adjustment algorithm couldn't handle hash-rate spikes.

**Comment by sebseb7** [comment]
"The hardfork we did was necessary"

doing it in secret gave me an additional advantage. Thank you again.

**Comment by sebseb7** [comment]
aritifically self limiting the official pool, also was advertisement for my zero fee pool - also thank you.

**Comment by aejontargaryen** [comment]
@sebseb7 this is cool. why can blocktime not be enforced?  because timestamps are changed?

**Comment by sebseb7** [comment]
"why can blocktime not be enforced?"

everyone needs to enforce them voluntary. Otherwise you only give an advantage to those not self limiting themselfes.

**Comment by nesterow** [comment]
@sebseb7 
" you only give an advantage to those not self limiting themselfes."
Seems like you have no idea what you talking about. If you restrict block time on BC level nodes won't accept blocks solved in less time, so your advantage would be sending orphan blocks to network.

**Comment by sebseb7** [comment]
this only breaks more stuff. Right now the official xun pools are wasting hashrate. And they lie to the miners. They don't tell my pool does not waste hashrate, they tell everybody my pool is stealing, when their pool does it. They even tried to fake a proov. That's very shady...

**Comment by sebseb7** [comment]
I have tons of screenshots proving they are criminal & incompetent.

**Comment by aejontargaryen** [comment]
interesting perspective.

**Comment by nesterow** [comment]
@sebseb7 Right now official pools work without limits. The limits were set for a week after we updated mainnet. IDK who tells everybody about your pool, I left it a link here because it is a good data-source. 
to track timewarps.

**Comment by sebseb7** [comment]
"Right now official pools work without limits."

When this is not a limit, is it stealing then?

https://i.imgur.com/lEDwyET.png

**Comment by sebseb7** [comment]
You accuse me of being some sort of criminal or cheater. When in reality Xun dev are criminal. I am just a miner.

**Comment by sebseb7** [comment]
I made my position clear and I will discuss this issue in public not in private. Everytime I discuss it in private out-of-context screenshots are made to accuse me of something... Iam done with those guys...

**Comment by nesterow** [comment]
@sebseb7  Lol. I don't accuse you in anything, after all this screenshot and link to your pool makes you some sort of "expert" on this matter.  
I put references here to discuss a technical issue - not to play troll games with you or anyone else. Тhe fact is that near to 100% of timewarped blocks are coming from your pool, it is out of any statistical abnormality.

**Comment by sebseb7** [comment]
" 100% of timewarped blocks are coming from your pool"

Because there is no other working pool. Official pools are broken. Average "luck" far above 100%

**Comment by sebseb7** [comment]
I delivered my blocks right to the dameon on the offical pool. ;-)

**Comment by nesterow** [comment]
Haha, I'll tell it to admins, that's a real story. lol. Did they leave daemon rpc on official pool wide open?

**Comment by sebseb7** [comment]
"Did they leave daemon rpc on official pool wide open?"

on all their hosts

**Comment by sebseb7** [comment]
I never even portscanned the wallet rpc port. Because my intention never was to steal or to cheat.

**Comment by sebseb7** [comment]
But you are probably part of the problem:

https://github.com/xun-project/UltraNote/issues/2

**Comment by nesterow** [comment]
@sebseb7 You could use blockchain explorer same way without port scanning, it uses same rpc port to get data. 

"I never even portscanned the wallet rpc port"
Wallet RPC should be running in a container listening on a bridge interface. It isn't accessible outside host.

"But you are probably part of the problem"
You didn't give any valid points in the issues you opened on xun repository. Only pointing to some config variables and trolling. I think I understand now why XUN team banned you everywhere possible. You are too obvious troll.

**Comment by sebseb7** [comment]
"Wallet RPC should be running in a container listening on a bridge interface"

tell that Xun devs, not me...

**Comment by FndNur1Labs** [comment]
update code with better hole security

**Comment by zawy12** [comment]
I've read the first post but not the rest. I wanted to make several comments.  

Karbowanec had to hard fork to replace the default Cryptonote/Monero difficulty that has an unacceptably long averaging window, N=720. The problem is that a small coin's price can change a lot in a day, but the difficulty needs N=720 (a day and a half on most of them) to catch up.  So a huge amount of mining comes on because the price/difficulty ratio is good. I saw 3x more mining in cases where difficulty was only 25% too low.  The converse also occurs. Difficulty maybe only 30% too high can see 70% of the miners leave until it drops.  This causes 3x delays in block finding. 

Sumokoin saw Karbowanec's sudden improvement when they switched to N=17 that I was promoting.  Compared to the default N=720, I think they were happy with it.  Masari however started having problems.  Actually they copied Sumokoin and the big problem was the N=17 was too low combined with Sumokoin using MTP as their more recent block so that negative solvetimes would not occur.  This is basically a 6-block delay in responding.  Simple moving average (SMA) algorithms see big mining jump on for about 1/3 of the difficulty window.  Sumo also used medians in a way that Masari claimed miners were taking advantage of, btu I don't think that made a big difference. Dynasty and Bytecoin appear to have also started the N=17.   

Sumokoin and Masari were allowing miners to get 6 blocks for "free" due to MTP and miners typically stay another 1/3 of the difficulty window (which is 6 blocks) while the difficulty begins to rise, then leave.  If they were 3x mining power, then 12 blocks later difficulty would be 2x what it was before they started.  And maybe with the higher difficulty,even more leave, so there were plenty of 3x target solvetimes when average over 10 or so blocks. With everyone gone, difficulty would plummet, and the cycle repeats.    Even the coins not using the delay would see this.

Instead of N=17, N=30 would have been better. Maybe even up to N=150 would have been better.  The ideal N is a balance between responding fast  and not accidentally varying too much.  It may be that N=12 is better if it responds so fast miners are not interested when it is accidentally too low.  But at N=12, it can easily be 50% too low.  The data I have is that N=60 is approximately best for most coins with a wide range of solvetimes. There's no data on N=12 but N=17 appears a lot worse than N=100.  So I've received a lot of praise from small coins, but now I see I made a mistake.  They would have done better with Digishield that uses a "tempering" with N=17 that causes it to act about like N=4x17= 68.  Even with the MTP delay, it's better than N=17 without the delay.  [Digishield has another problem](https://github.com/zawy12/difficulty-algorithms/issues/9).

Adjusting reward based on difficulty seemed promising, but it has the same old problem as a low N. Miners are motivated by price/difficulty ratio. Changing the reward is like changing the price. It's trying to motivate or dis-motivate by changing the numerator in addition to the denominator.  It's like "doubly" changing the denominator.  It turns out, yes it works, but it has the same old problem as low N that changes difficulty "doubly" fast:  the ratio would also "doubly" accidentally go too low too often to proportion to its speed. So it was attracting "hash attacks" by accidentally going too low in the same proportion as it was dis-motivating them by responding quickly.

"Timewarp" as originally described in a bitcointalk post years ago is a specific type of attack that can't happen with rolling averages.  Miners coming on and off suddenly is not a timestamp-related attack.   Timestamps can be changed to affect difficulty and there is no fix. This limits how good difficulty algorithms can be made.  It's central to Satoshi's ideas that miners select the timestamps.  It's also central for BTC to work that nodes limit how far ahead or behind the timestamps can be,

The [LWMA aka WHM algorithm](https://github.com/zawy12/difficulty-algorithms/issues/3) is the best.  I would just use it and not worry about difficulty again, unless there's a major trend showing an improvement.  But please use a solvetime variable that allows negative solvetimes as a way to avoid the delays caused by MTP.  Follow the algorithm precisely.  I know 2 coins that left a catastrophic exploit by not following it exactly. I then have to email them to describe the exploit and they have to fork again.  It's mathematically correct and logical to allow the negative solvetimes.  It's not logically correct to reject negative solvetimes simply based on the idea "time can't be negative".

Andrew Stone has an idea similar to changing reward based on difficulty.  His idea is to increase reward if solvetime goes past some limit.  But it's a major change in difficulty and possibly not trivial to implement. I think the LWMA is good enough that we don't need to worry about difficulty anymore.

**Comment by nesterow** [comment]
@zawy12 Thank you for the comment and useful information. I'm sorry I got confused at first who were the algorithms' authors.
I think that penalizing reward based on timestamps was a bad idea. Simply because there's no way to know if a timestamp was manipulated. And it is not fare for miners. Andrew Stone's idea seems more fare from miner's perspective.
Enforcing limits on block solving time (not allowing some negative time threshold) maybe only solution to protect from timestamp manipulations. But it requires heavy modifications for most things related to mining and I doubt it would be flawless. Not only because of reasons you mentioned, but because it affects almost every part of PoW.
As I learn more things I get to idea that for small coins using PoW may be not a good idea.

**Comment by zawy12** [comment]
I mis-stated [Andrew's method](url)https://medium.com/@g.andrew.stone/the-blockchain-difficulty-information-paradox-879b0336864f.  His does not change reward, it decreases difficulty as the solvetime goes further past the target solvetime.  

BTC has timestamp limits.  The forward limit of 7200 seconds past median of node peers is based on thinking 12 block ahead was "too far".    A faster coin needs a similar reduction in the limit, down to 12 block or less.   So a 120 second coin needs it changed down to 1440 seconds, or less.  I believe Monero and/or crypto might have a limit of 24 blocks which is a mistake.  My algorithms have a limit of 6 or 7 blocks based on previous timestamp, which is a different type of limit than the one based on median of node peer real time.  The limit based on node time is inherently needed.  The first post above has a link where I try to cover everything about bad timestamps.  The best solution I've found is to limit each timestamp to +6x to -5x the target solvetime of the previous timestamp.  If it were not for bad timestamps, an improvement to algorithms could be made: they could shift to a smaller averaging window if the last 5 or so solvetimes were statistically different from the target.  For complicated reasons having to do with the effect of bad timestamps, it has a questionable benefit at added complexity.

**Comment by nesterow** [comment]
Yes, Cryptonote derivatives use 7200 seconds past median as well. Monero uses window of 60 past blocks.  I can see a possible problem here. Maybe I got the idea wrong, but node assigns block template with timestamp taken from node's time. So if forward time limit is too small it can happen that all nodes end up with invalid block candidates. I know It is a ridiculously low possibility, but still it can be the case. And if I got the concept right - less the forward limit more possibility for this scenario to happen.

**Comment by zawy12** [comment]
You mean Cryptonote validaets a block if the timestamp is not more than 7200 seconds ahead of median node time?  Miners can own local node(s).  This was part of the problem in timewarp attacks.  In BTC the timestamp can't be older than the median of past 11 blocks, and not newer than 7200 seconds ahead of the median of local nodes.   Zcash regretted using 3600 instead of 1500 seconds because they have a target time of 150 seconds.  If Cryptonote's defaulting to 7200 seconds ahead of node time, then one only needs to assign that max time to drive difficulty down 7200/T/ N = 53% in 1 block for T = target solvetime = 150 and N = 90.  If the coin also is not allowing negative solvetimes, then blocks with the correct time are not going to be able to counteract the bad timestamp, so it's going to take 90 blocks for the difficulty to rise back up to where it should be (if it's a SMA).  MTP is used as the most recent block to prevent this, but the +6 / -5 limit from previous timestamp works better.  A miner with 50% can "own" the MTP and still drive difficulty down thanks to that overly generous 7200 limit.

**Comment by nesterow** [comment]
> You mean Cryptonote validaets a block if the timestamp is not more than 7200 seconds ahead of median node time? 

[It appears so](https://github.com/monero-project/monero/blob/ed67e5c001d50a2eea5fa24dfc58eba0921d525b/src/cryptonote_core/blockchain.cpp#L3101). I can be wrong on this one. Too much code to inspect to be sure.

> A miner with 50% can "own" the MTP and still drive difficulty down 

That's was the first reason why I had doubts about enforcing constant time limits on blocks. It messes with everything. It won't adjust difficulty properly in any direction. And mining would have less sense: At low difficulty miners would take most reward, at high difficulty they would stall to drop difficulty back. And it would add more problems in different parts of PoW workflow - block templates, pools, mining software, etc.
